network/p2p/unicast/manager.go

```diff
@@ -138,49 +244,108 @@ func (m *Manager) rawStreamWithProtocol(ctx context.Context,
 		// cancel the dial back off (if any), since we want to connect immediately
 		dialAddr = m.streamFactory.DialAddress(peerID)
 		m.streamFactory.ClearBackoff(peerID)
-
-		// if this is a retry attempt, wait for some time before retrying
-		if retries > 0 {
-			// choose a random interval between 0 to 5
-			// (to ensure that this node and the target node don't attempt to reconnect at the same time)
-			r := rand.Intn(MaxConnectAttemptSleepDuration)
-			time.Sleep(time.Duration(r) * time.Millisecond)
-		}
-
 		err := m.streamFactory.Connect(ctx, peer.AddrInfo{ID: peerID})
 		if err != nil {
-
 			// if the connection was rejected due to invalid node id, skip the re-attempt
 			if strings.Contains(err.Error(), "failed to negotiate security protocol") {
-				return s, dialAddr, fmt.Errorf("invalid node id: %w", err)
+				return fmt.Errorf("failed to dial remote peer: %w", err)
 			}
 
 			// if the connection was rejected due to allowlisting, skip the re-attempt
 			if errors.Is(err, swarm.ErrGaterDisallowedConnection) {
-				return s, dialAddr, fmt.Errorf("target node is not on the approved list of nodes: %w", err)
+				return fmt.Errorf("target node is not on the approved list of nodes: %w", err)
 			}
 
-			errs = multierror.Append(errs, err)
-			continue
+			m.logger.Warn().
+				Err(err).
+				Str("peer_id", peerID.String()).
+				Int("attempt", dialAttempts).
+				Uint64("max_attempts", maxAttempts).
+				Msg("retrying peer dialing")
+
+			return retry.RetryableError(err)
+		}
+
+		return nil
+	}
+
+	// retryable func that will attempt to create the stream using the stream factory if connection exists
+	createStreamAttempts := 0
+	createStream := func(context.Context) error {
+		createStreamAttempts++
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context done before stream could be created (retry attempt: %d, errors: %w)", createStreamAttempts, errs)
+		default:
 		}
 
+		var err error
+		// add libp2p context value NoDial to prevent the underlying host from dialingComplete the peer while creating the stream
+		// we've already ensured that a connection already exists.
+		ctx = libp2pnet.WithNoDial(ctx, "application ensured connection to peer exists")
 		// creates stream using stream factory
 		s, err = m.streamFactory.NewStream(ctx, peerID, protocolID)
 		if err != nil {
 			// if the stream creation failed due to invalid protocol id, skip the re-attempt
 			if strings.Contains(err.Error(), "protocol not supported") {
```

Comment:
similarly to my previous comment, also this error inspection should be moved to the `StreamFactory` wrapper around libP2P

Reply:
https://github.com/onflow/flow-go/pull/3902/commits/9452b35c05dbfc7b2b0c50aeb0b652200e845dae